eahub/base/views.py

```diff
@@ -94,28 +94,77 @@ def groups(request):
     )
 
 
-def get_groups_data():
-    rows = Group.objects.all()
+def get_groups_data(user, by_distance=False):
+    if hasattr(user, "profile") and by_distance:
+        if hasattr(user.profile, "lat"):
+            rows = get_groups_by_distance(user.profile)
+    else:
+        rows = Group.objects.all()
     map_data = [
         {
-            "lat": x.lat,
-            "lng": x.lon,
-            "label": x.name,
-            "active": x.is_active,
-            "path": f"/group/{x.slug}",
+            "lat": group.lat,
+            "lng": group.lon,
+            "label": group.name,
+            "active": group.is_active,
+            "path": f"/group/{group.slug}",
         }
-        for x in rows
-        if x.lat and x.lon
+        for group in rows
+        if group.lat and group.lon
     ]
     return {"rows": rows, "map_data": map_data}
 
 
+def get_groups_by_distance(profile):
+    # Difference in latitude (multiplied by 2 because latitude ranges only to 90)
+    # plus difference in longitude (databases cannot calculate square roots so no
+    # Pythagorean theorem)
+    equalized_user_lat = profile.lat + 90
+    equalized_user_lon = (profile.lon + 180) / 2
+    return (
+        Group.objects.annotate(equalized_lat=F("lat") + 90)
+        .annotate(zero_centered_lon=F("lon") + 180)
+        .annotate(equalized_lon=F("zero_centered_lon") / 2)
+        .annotate(
+            lat_difference=Case(
+                When(
+                    equalized_lat__lte=equalized_user_lat,
+                    then=equalized_user_lat - F("equalized_lat"),
+                ),
+                When(
+                    equalized_lat__gt=equalized_user_lat,
+                    then=F("equalized_lat") - equalized_user_lat,
+                ),
+            )
+        )
+        .annotate(
+            lon_difference=Case(
+                When(
+                    equalized_lon__lte=equalized_user_lon,
+                    then=equalized_user_lon - F("equalized_lon"),
+                ),
+                When(
+                    equalized_lon__gt=equalized_user_lon,
+                    then=F("equalized_lon") - equalized_user_lon,
+                ),
+            )
+        )
+        .annotate(total_difference=F("lat_difference") + F("lon_difference"))
+        .order_by("total_difference")
+        .all()
+    )
+
+
 def get_profiles_data(user):
     rows = Profile.objects.visible_to_user(user)
     map_data = [
-        {"lat": x.lat, "lng": x.lon, "label": x.name, "path": f"/profile/{x.slug}"}
-        for x in rows
-        if x.lat and x.lon
+        {
+            "lat": profile.lat,
+            "lng": profile.lon,
+            "label": profile.name,
+            "path": f"/profile/{profile.slug}",
```

Comment:
This should be using Django's `reverse` function instead of manually constructing the URL path.